assembly/src/parser/error.rs

```diff
@@ -0,0 +1,312 @@
+use alloc::{
+    string::{String, ToString},
+    vec::Vec,
+};
+use core::{fmt, ops::Range};
+
+use super::{ParseError, SourceSpan};
+use crate::diagnostics::Diagnostic;
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum LiteralErrorKind {
+    /// The input was empty
+    Empty,
+    /// The input contained an invalid digit
+    InvalidDigit,
+    /// The value overflows `u32::MAX`
+    U32Overflow,
+    /// The value overflows `Felt::MODULUS`
+    FeltOverflow,
+    /// The value was expected to be a power of two, and was not
+    InvalidPowerOfTwo,
+}
+impl fmt::Display for LiteralErrorKind {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Self::Empty => f.write_str("input was empty"),
+            Self::InvalidDigit => f.write_str("invalid digit"),
+            Self::U32Overflow => f.write_str("value overflowed the u32 range"),
+            Self::FeltOverflow => f.write_str("value overflowed the field modulus"),
+            Self::InvalidPowerOfTwo => {
+                f.write_str("expected value to be a power of two, but it wasn't")
+            }
+        }
+    }
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum HexErrorKind {
+    /// Expected two hex digits for every byte, but had fewer than that
+    MissingDigits,
+    /// Valid hex-encoded integers are expected to come in sizes of 8, 16, or 64 digits,
+    /// but the input consisted of an invalid number of digits.
+    Invalid,
+    /// Occurs when a hex-encoded value overflows `Felt::MODULUS`, the maximum integral value
+    Overflow,
+    /// Occurs when the hex-encoded value is > 64 digits
+    TooLong,
+}
+impl fmt::Display for HexErrorKind {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Self::MissingDigits => {
+                f.write_str("expected number of hex digits to be a multiple of 2")
+            }
+            Self::Invalid => f.write_str("expected 2, 4, 8, 16, or 64 hex digits"),
+            Self::Overflow => f.write_str("value overflowed the field modulus"),
+            Self::TooLong => f.write_str(
+                "value has too many digits, long hex strings must contain exactly 64 digits",
+            ),
+        }
+    }
+}
+
+#[derive(Debug, Default, Clone, thiserror::Error, Diagnostic)]
+#[repr(u8)]
+pub enum ParsingError {
+    #[default]
+    #[error("parsing failed due to unexpected input")]
+    #[diagnostic()]
+    Failed = 0,
+    #[error("invalid token")]
+    #[diagnostic()]
+    InvalidToken {
+        #[label("occurs here")]
+        span: SourceSpan,
+    },
+    #[error("unrecognized token")]
+    #[diagnostic(help("expected {}", expected.as_slice().join(", or ")))]
+    UnrecognizedToken {
+        #[label("lexed a {token} here")]
```

Comment:
```suggestion
        #[label("found a {token} here")]
```

I wouldn't use compiler terminology in user-facing errors/diagnostics

Reply:
Good point, that'll hit a bunch of tests, but it's basically a simple find/replace, so not a big deal